File: 家谱树.cpp
#include <bits/stdc++.h>
using namespace std;

#define int long long

signed main() {
    int n, x;
    cin >> n;
    vector<int> g[n + 1];
    int cnt[n + 1];
    memset(cnt, 0, sizeof(cnt));
    for (int i = 1; i <= n; ++i) {
        while (true) {
            cin >> x;
            if (!x) {
                break;
            }
            g[i].emplace_back(x);
            cnt[x]++;
        }
    }

    queue<int> q;
    for (int i = 1; i <= n; ++i) {
        if (!cnt[i]) {
            q.push(i);
        }
    }
    while (q.size()) {
        int u = q.front();
        q.pop();
        cout << u << " ";
        for (int v : g[u]) {
            if (--cnt[v] == 0) {
                q.push(v);
            } 
        }
    }
    return 0;
}

File: 家谱树.py
from queue import Queue

n = int(input())
g, cnt = [[] for _ in range(n + 1)], [0 for _ in range(n + 1)]
for i in range(1, n + 1):
    a = list(map(int, input().split()))
    for x in a[:-1]:
        g[i].append(x)
        cnt[x] += 1

q = Queue()
for i in range(1, n + 1):
    if not cnt[i]:
        q.put(i)
while q.qsize():
    u = q.get()
    print(u, end = " ")
    for v in g[u]:
        cnt[v] -= 1
        if not cnt[v]:
            q.put(v)
